worker/raft/apply.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2021 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package raft
+
+import (
+	"time"
+
+	"github.com/hashicorp/raft"
+	"github.com/juju/clock"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/raftlease"
+	"github.com/juju/juju/worker/raft/queue"
+)
+
+// Raft defines a local use Raft instance.
+type Raft interface {
+	// State is used to return the current raft state.
+	State() raft.RaftState
+
+	// Leader is used to return the current leader of the cluster.
+	// It may return empty string if there is no current leader
+	// or the leader is unknown.
+	Leader() raft.ServerAddress
+
+	// GetConfiguration returns the latest configuration. This may not yet be
+	// committed. The main loop can access this directly.
+	GetConfiguration() raft.ConfigurationFuture
+
+	// Apply is used to apply a command to the FSM in a highly consistent
+	// manner. This returns a future that can be used to wait on the application.
+	// An optional timeout can be provided to limit the amount of time we wait
+	// for the command to be started. This must be run on the leader or it
+	// will fail.
+	Apply([]byte, time.Duration) raft.ApplyFuture
+}
+
+// Applier applies a new operation against a raft instance.
+type Applier struct {
+	raft         Raft
+	notifyTarget raftlease.NotifyTarget
+	clock        clock.Clock
+	logger       Logger
+}
+
+// NewApplier creates a new Applier.
+func NewApplier(raft Raft, target raftlease.NotifyTarget, clock clock.Clock, logger Logger) LeaseApplier {
+	return &Applier{
+		raft:         raft,
+		notifyTarget: target,
+		clock:        clock,
+		logger:       logger,
+	}
+}
+
+// ApplyOperation applies an lease opeartion against the raft instance. If the
+// raft instance isn't the leader, then an error is returned with the leader
+// information if available.
+// This Raft spec outlines this "The first option, which we recommend ..., is
+// for the server to reject the request and return to the client the address of
+// the leader, if known." (see 6.2.1).
+// If the leader is the current raft instance, then attempt to apply it to
+// the fsm.
+func (a *Applier) ApplyOperation(op queue.Operation) error {
+	if state := a.raft.State(); state != raft.Leader {
+		leaderAddress := a.raft.Leader()
+
+		a.logger.Infof("Attempt to apply the lease failed, we're not the leader. State: %v, Leader: %v", state, leaderAddress)
```

Comment:
Should it be info-level? Maybe. If the leader isn't jumping about we shouldn't get this too often.

Reply:
This might happen a lot when starting up, but once settled it should be an indicator that the leader is flapping and will cause issues with the lease manager. So this is the canary in the coal mine.